src/rosdiscover/recover/core.py

```diff
@@ -0,0 +1,257 @@
+# -*- coding: utf-8 -*-
+"""
+This module provides several data structures that are used to describe
+architectural models that have been recovered from nodes.
+"""
+__all__ = ('PublisherDefinition',
+           'ParameterDelete',
+           'ParameterExistence',
+           'ParameterRead',
+           'ParameterSearch',
+           'ParameterWrite',
+           'RecoveredNodeModel',
+           'RecoveredNodeModelElement',
+           'ServiceDefinition',
+           'ServiceProxyDefinition',
+           'SimpleActionClientDefinition',
+           'SimpleActionServerDefinition',
+           'SubscriberDefinition')
+
+from typing import Any, Collection, Optional
+import abc
+
+import attr
+import sourcelocation
+
+
+class RecoveredNodeModelElement(abc.ABC):
+    """Represents an element of a recovered node model.
+
+    Attributes
+    ----------
+    source: str
+        The associated source code for this element.
+    location: sourcelocation.FileLocationRange
+        The location of the source code for the element.
+    """
+    @abc.abstractmethod
+    @property
+    def source(self) -> str:
+        ...
+
+    @abc.abstractmethod
+    @property
+    def location(self) -> sourcelocation.FileLocationRange:
+        ...
+
+
+@attr.s(slots=True, frozen=True, auto_attribs=True)
+class ServiceDefinition(RecoveredNodeModelElement):
+    """Describes the creation of a service.
+
+    Attributes
+    ----------
+    type_: str
+        The name of the type used by the service.
+    service: str
+        The name of the service.
+    """
+    source: str
+    location: sourcelocation.FileLocationRange
+    type_: str
+    service: str
+
+
+@attr.s(slots=True, frozen=True, auto_attribs=True)
+class ServiceProxyDefinition(RecoveredNodeModelElement):
```

Comment:
Is a server proxy the same as a service client? Perhaps "ServiceClient" would be a better name?